File: Windows Source/TotallyHuman/Models/SharedTypes.cs
using System;
using System.Collections.Generic;
using System.Text.Json.Serialization;

namespace TotallyHuman.Models;

// SharedTypes.cs — Gemeinsame Datenmodelle (1:1 Port von SharedTypes.swift)
// Totally Human — KI-Musik-Detektor / AI Music Detector

/// <summary>Zentrale Analyse-Konfiguration.</summary>
public static class AnalysisConfig
{
    public const double StandardSampleRate = 44100.0;
    public static readonly int[] FftSizes = { 2048, 4096, 8192 };
    public static readonly double[] FftWeights = { 0.25, 0.45, 0.30 };
    public const double SsmWeight = 0.15;
    public static readonly (double Unten, double Oben) AnalysisFreqBandHz = (5000.0, 16000.0);
    public const double SegmentLengthSeconds = 10.0;
    public const int FeatureVectorSize = 128;
    public static readonly HashSet<string> SupportedExtensions = new(StringComparer.OrdinalIgnoreCase)
        { "mp3", "wav", "aiff", "aif", "m4a", "flac", "ogg", "opus", "wma", "alac" };
}

// MARK: - Artefakt-Ergebnis (Fingerprint einer FFT-Größe)
public struct ArtefaktErgebnis
{
    public double[] Frequenzen;
    public double[] MittleresSpektrumDB;
    public double[] GrundlinieDB;
    public double[] Fingerprint;
    public int[] PeakIndizes;
    public double[] PeakFrequenzen;
    public double[] PeakProminenzen;
    public double PeakAbstandHz;
    public double Regelmaessigkeit;
    public double ArtefaktStaerke;
    public int AnzahlPeaks;
}

// MARK: - Ensemble-Ergebnis
public struct EnsembleErgebnis
{
    public ArtefaktErgebnis HauptErgebnis;
    public double[] EinzelScores;
    public double EnsembleVarianz;
    public double EnsembleScore;
}

/// <summary>Empfehlungs-Kategorie der Detektion.</summary>
public enum Empfehlung
{
    KlarKiGeneriert,
    KlarMenschlich,
    ManuellePruefung,
    Verschleierungsverdacht
}

public static class EmpfehlungExtensions
{
    /// <summary>Roher String-Wert (kompatibel mit Swift-Enum rawValue).</summary>
    public static string RawValue(this Empfehlung e) => e switch
    {
        Empfehlung.KlarKiGeneriert => "KLAR KI-GENERIERT",
        Empfehlung.KlarMenschlich => "KLAR MENSCHLICH",
        Empfehlung.ManuellePruefung => "MANUELLE PRÜFUNG EMPFOHLEN",
        Empfehlung.Verschleierungsverdacht => "VERSCHLEIERUNGSVERDACHT",
        _ => ""
    };

    /// <summary>Farbe (red/green/orange/purple) wie im Original.</summary>
    public static string Farbe(this Empfehlung e) => e switch
    {
        Empfehlung.KlarKiGeneriert => "red",
        Empfehlung.KlarMenschlich => "green",
        Empfehlung.ManuellePruefung => "orange",
        Empfehlung.Verschleierungsverdacht => "purple",
        _ => "orange"
    };
}

/// <summary>Analysemodus: heuristisch oder trainiert.</summary>
public enum AnalyseModus
{
    Heuristisch,
    Trainiert
}

public static class AnalyseModusExtensions
{
    public static string RawValue(this AnalyseModus m) => m == AnalyseModus.Trainiert ? "trainiert" : "heuristisch";
}

/// <summary>Vollständiges Detektions-Ergebnis (1:1 zu DetektionsErgebnis.swift).</summary>
public class DetektionsErgebnis
{
    public Guid Id { get; set; } = Guid.NewGuid();
    public string DateiPfad { get; set; } = "";
    public string DateiName { get; set; } = "";
    public bool IstKIGeneriert { get; set; }
    public double KiWahrscheinlichkeitProzent { get; set; }
    public double ArtefaktScore { get; set; }
    public string Einstufung { get; set; } = "";
    public Empfehlung Empfehlung { get; set; } = Empfehlung.ManuellePruefung;
    public double KonfidenzScore { get; set; }
    public string KonfidenzText { get; set; } = "";
    public int AnzahlPeaks { get; set; }
    public double MittlererPeakAbstandHz { get; set; }
    public double PeakRegelmaessigkeit { get; set; }
    public double SsmScore { get; set; }
    public double[] EnsembleScores { get; set; } = Array.Empty<double>();
    public double EnsembleVarianz { get; set; }
    public double ZeitlicherKIAnteilProzent { get; set; }
    public int[] SplicePositionen { get; set; } = Array.Empty<int>();
    public List<string> VerschleierungsHinweise { get; set; } = new();
    public bool VerschleierungsVerdacht { get; set; }
    public AnalyseModus Modus { get; set; } = AnalyseModus.Heuristisch;
    public List<string> Hinweise { get; set; } = new();
    public DateTime AnalyseDatum { get; set; } = DateTime.Now;

    // Detail-Daten für die Visualisierung
    public double[] FrequenzBandAnteile { get; set; } = Array.Empty<double>();
    public double[][] FrequenzBandGrenzen { get; set; } = Array.Empty<double[]>();
    public double[][] SegmentZeiten { get; set; } = Array.Empty<double[]>();
    public double[] SegmentStaerken { get; set; } = Array.Empty<double>();
    public double[] SpektrumFrequenzen { get; set; } = Array.Empty<double>();
    public double[] SpektrumWerte { get; set; } = Array.Empty<double>();
    public double[] FingerprintWerte { get; set; } = Array.Empty<double>();
    public double[] GrundlinieWerte { get; set; } = Array.Empty<double>();
    public double[][] SsmMatrix { get; set; } = Array.Empty<double[]>();
}

/// <summary>Label für Trainingsproben.</summary>
public enum TrainingsLabel
{
    Echt = 0,
    Ki = 1
}

/// <summary>Eine einzelne Trainingsprobe (128-dim Merkmalsvektor).</summary>
public class TrainingsProbe
{
    [JsonPropertyName("id")]
    public Guid Id { get; set; } = Guid.NewGuid();

    [JsonPropertyName("dateiPfad")]
    public string DateiPfad { get; set; } = "";

    [JsonPropertyName("dateiName")]
    public string DateiName { get; set; } = "";

    // Serialisiert als Int (0/1) — kompatibel mit Swift & Seed-JSON.
    [JsonPropertyName("label")]
    public int LabelRaw { get; set; }

    [JsonIgnore]
    public TrainingsLabel Label
    {
        get => LabelRaw == 1 ? TrainingsLabel.Ki : TrainingsLabel.Echt;
        set => LabelRaw = (int)value;
    }

    [JsonPropertyName("merkmalsvektor")]
    public double[] Merkmalsvektor { get; set; } = Array.Empty<double>();

    [JsonPropertyName("dateiSignatur")]
    public string DateiSignatur { get; set; } = "";

    [JsonPropertyName("hinzugefuegtAm")]
    public DateTime HinzugefuegtAm { get; set; } = DateTime.Now;
}

/// <summary>Logistisches Modell (Gewichte + Bias). 1:1 zu LogistischesModell.swift.</summary>
public class LogistischesModell
{
    [JsonPropertyName("gewichte")]
    public double[] Gewichte { get; set; } = Array.Empty<double>();

    [JsonPropertyName("bias")]
    public double Bias { get; set; }

    [JsonPropertyName("trainingsZeit")]
    public DateTime TrainingsZeit { get; set; } = DateTime.Now;

    [JsonPropertyName("anzahlBeispiele")]
    public int AnzahlBeispiele { get; set; }

    [JsonPropertyName("kreuzvalidierungsGenauigkeit")]
    public double KreuzvalidierungsGenauigkeit { get; set; }

    /// <summary>Sigmoid-Vorhersage der KI-Wahrscheinlichkeit.</summary>
    public double Vorhersage(double[] merkmale)
    {
        if (merkmale.Length != Gewichte.Length) return 0.5;
        double summe = Bias;
        for (int i = 0; i < merkmale.Length; i++)
            summe += Gewichte[i] * merkmale[i];
        return Sigmoid(summe);
    }

    private static double Sigmoid(double x)
    {
        if (x >= 0) return 1.0 / (1.0 + Math.Exp(-x));
        double z = Math.Exp(x);
        return z / (1.0 + z);
    }
}

File: macOS Source/TotallyHuman/Models/SharedTypes.swift
// SharedTypes.swift — Gemeinsame Datenmodelle für die gesamte App
// Totally Human — KI-Musik-Detektor (native macOS, kein Python)
// Alle Module referenzieren diese Typen.

import Foundation

// MARK: - Analyse-Konfiguration
struct AnalysisConfig {
    static let standardSampleRate: Double = 44100.0
    static let fftSizes: [Int] = [2048, 4096, 8192]
    static let fftWeights: [Double] = [0.25, 0.45, 0.30]
    static let ssmWeight: Double = 0.15
    static let analysisFreqBandHz: (Double, Double) = (5000.0, 16000.0)
    static let segmentLengthSeconds: Double = 10.0
    static let featureVectorSize: Int = 128
    static let supportedExtensions: Set<String> = ["mp3","wav","aiff","aif","m4a","flac","ogg","opus","wma","alac"]
}

// MARK: - Artefakt-Ergebnis (Fingerprint einer FFT-Größe)
struct ArtefaktErgebnis {
    var frequenzen: [Double]           // Hz-Werte für jedes Spektrum-Bin
    var mittleresSpektrumDB: [Double]  // Gemitteltes Amplitudenspektrum in dB
    var grundlinieDB: [Double]         // Morphologisch geschätzte Grundlinie
    var fingerprint: [Double]          // Spektrum minus Grundlinie
    var peakIndizes: [Int]             // Indizes der erkannten Peaks
    var peakFrequenzen: [Double]       // Hz-Werte der Peaks
    var peakProminenzen: [Double]      // Prominenz jedes Peaks in dB
    var peakAbstandHz: Double          // Mittlerer Peak-Abstand
    var regelmaessigkeit: Double       // Regelmäßigkeit 0..1
    var artefaktStaerke: Double        // Haupt-Score
    var anzahlPeaks: Int               // Anzahl erkannter Peaks
}

// MARK: - Ensemble-Ergebnis
struct EnsembleErgebnis {
    var hauptErgebnis: ArtefaktErgebnis  // Repräsentativ für 4096-FFT
    var einzelScores: [Double]          // Score je FFT-Größe
    var ensembleVarianz: Double          // Varianz (hoch = Manipulationsverdacht)
    var ensembleScore: Double            // Gewichtetes Mittel
}

// MARK: - Spektrale Merkmale
struct SpektralMerkmale {
    var hochfrequenzAnteil: Double        // Energie > 8 kHz
    var spektraleFlachheit: Double        // Flatness [0..1]
    var cutoffSchaerfe: Double            // Schärfe eines HF-Cutoffs
    var zeitlicheStationaritaet: Double   // Temporale Stationarität [0..1]
}

// MARK: - Frequenzband-Ergebnis
struct FrequenzBandErgebnis {
    var bandGrenzenHz: [(Double, Double)]  // (Untergrenze, Obergrenze) je Band
    var bandStaerken: [Double]             // Artefaktstärke je Band
    var bandAnteile: [Double]              // Prozentualer Anteil je Band
    var staerktesBandHz: (Double, Double) // Stärkstes Band
}

// MARK: - Zeit-Segment-Ergebnis
struct ZeitSegmentErgebnis {
    var segmentZeitenS: [(Double, Double)]  // (Start, Ende) in Sekunden
    var segmentStaerken: [Double]           // Artefaktstärke je Segment
    var segmentAnteile: [Double]            // Prozentualer Anteil
    var staerktesSegmentS: (Double, Double) // Stärkstes Segment
}

// MARK: - Splice-Erkennung
struct SpliceErgebnis {
    var spliceVerdacht: Bool
    var splicePositionen: [Int]
    var maxGradient: Double
}

// MARK: - Verschleierungs-Analyse
struct VerschleierungsErgebnis {
    var verdachtGesamt: Bool
    var hinweise: [String]
    var spliceVerdacht: Bool
    var masteringAnomalie: Bool
    var wiederholungsAnomalie: Bool
}

// MARK: - Vollständiges Detektions-Ergebnis
struct DetektionsErgebnis: Identifiable, Codable {
    let id: UUID
    var dateiPfad: String
    var dateiName: String
    var istKIGeneriert: Bool
    var kiWahrscheinlichkeitProzent: Double
    var artefaktScore: Double
    var einstufung: String
    var empfehlung: Empfehlung
    var konfidenzScore: Double
    var konfidenzText: String
    var anzahlPeaks: Int
    var mittlererPeakAbstandHz: Double
    var peakRegelmaessigkeit: Double
    var ssmScore: Double
    var ensembleScores: [Double]
    var ensembleVarianz: Double
    var zeitlicherKIAnteilProzent: Double
    var splicePositionen: [Int]
    var verschleierungsHinweise: [String]
    var verschleierungsVerdacht: Bool
    var modus: AnalyseModus
    var hinweise: [String]
    var analyseDatum: Date
    
    // Optionale Detail-Daten (für Visualisierung)
    var frequenzBandAnteile: [Double]
    var frequenzBandGrenzen: [[Double]]
    var segmentZeiten: [[Double]]
    var segmentStaerken: [Double]
    var spektrumFrequenzen: [Double]
    var spektrumWerte: [Double]
    var fingerprintWerte: [Double]
    var grundlinieWerte: [Double]
    var ssmMatrix: [[Double]]
    
    enum Empfehlung: String, Codable, CaseIterable {
        case klarkiGeneriert = "KLAR KI-GENERIERT"
        case klarMenschlich = "KLAR MENSCHLICH"
        case manuellePruefung = "MANUELLE PRÜFUNG EMPFOHLEN"
        case verschleierungsverdacht = "VERSCHLEIERUNGSVERDACHT"
        
        var farbe: String {
            switch self {
            case .klarkiGeneriert: return "red"
            case .klarMenschlich: return "green"
            case .manuellePruefung: return "orange"
            case .verschleierungsverdacht: return "purple"
            }
        }
        
        var symbol: String {
            switch self {
            case .klarkiGeneriert: return "xmark.circle.fill"
            case .klarMenschlich: return "checkmark.circle.fill"
            case .manuellePruefung: return "questionmark.circle.fill"
            case .verschleierungsverdacht: return "exclamationmark.triangle.fill"
            }
        }
    }
    
    enum AnalyseModus: String, Codable {
        case heuristisch = "heuristisch"
        case trainiert = "trainiert"
    }
    
    // Vollständiger Initialisierer (id wird automatisch erzeugt).
    init(
        dateiPfad: String,
        dateiName: String,
        istKIGeneriert: Bool,
        kiWahrscheinlichkeitProzent: Double,
        artefaktScore: Double,
        einstufung: String,
        empfehlung: Empfehlung,
        konfidenzScore: Double,
        konfidenzText: String,
        anzahlPeaks: Int,
        mittlererPeakAbstandHz: Double,
        peakRegelmaessigkeit: Double,
        ssmScore: Double,
        ensembleScores: [Double],
        ensembleVarianz: Double,
        zeitlicherKIAnteilProzent: Double,
        splicePositionen: [Int],
        verschleierungsHinweise: [String],
        verschleierungsVerdacht: Bool,
        modus: AnalyseModus,
        hinweise: [String],
        analyseDatum: Date,
        frequenzBandAnteile: [Double],
        frequenzBandGrenzen: [[Double]],
        segmentZeiten: [[Double]],
        segmentStaerken: [Double],
        spektrumFrequenzen: [Double],
        spektrumWerte: [Double],
        fingerprintWerte: [Double],
        grundlinieWerte: [Double],
        ssmMatrix: [[Double]]
    ) {
        self.id = UUID()
        self.dateiPfad = dateiPfad
        self.dateiName = dateiName
        self.istKIGeneriert = istKIGeneriert
        self.kiWahrscheinlichkeitProzent = kiWahrscheinlichkeitProzent
        self.artefaktScore = artefaktScore
        self.einstufung = einstufung
        self.empfehlung = empfehlung
        self.konfidenzScore = konfidenzScore
        self.konfidenzText = konfidenzText
        self.anzahlPeaks = anzahlPeaks
        self.mittlererPeakAbstandHz = mittlererPeakAbstandHz
        self.peakRegelmaessigkeit = peakRegelmaessigkeit
        self.ssmScore = ssmScore
        self.ensembleScores = ensembleScores
        self.ensembleVarianz = ensembleVarianz
        self.zeitlicherKIAnteilProzent = zeitlicherKIAnteilProzent
        self.splicePositionen = splicePositionen
        self.verschleierungsHinweise = verschleierungsHinweise
        self.verschleierungsVerdacht = verschleierungsVerdacht
        self.modus = modus
        self.hinweise = hinweise
        self.analyseDatum = analyseDatum
        self.frequenzBandAnteile = frequenzBandAnteile
        self.frequenzBandGrenzen = frequenzBandGrenzen
        self.segmentZeiten = segmentZeiten
        self.segmentStaerken = segmentStaerken
        self.spektrumFrequenzen = spektrumFrequenzen
        self.spektrumWerte = spektrumWerte
        self.fingerprintWerte = fingerprintWerte
        self.grundlinieWerte = grundlinieWerte
        self.ssmMatrix = ssmMatrix
    }

    init(dateiPfad: String) {
        self.id = UUID()
        self.dateiPfad = dateiPfad
        self.dateiName = URL(fileURLWithPath: dateiPfad).lastPathComponent
        self.istKIGeneriert = false
        self.kiWahrscheinlichkeitProzent = 0
        self.artefaktScore = 0
        self.einstufung = ""
        self.empfehlung = .manuellePruefung
        self.konfidenzScore = 0
        self.konfidenzText = ""
        self.anzahlPeaks = 0
        self.mittlererPeakAbstandHz = 0
        self.peakRegelmaessigkeit = 0
        self.ssmScore = 0
        self.ensembleScores = []
        self.ensembleVarianz = 0
        self.zeitlicherKIAnteilProzent = 0
        self.splicePositionen = []
        self.verschleierungsHinweise = []
        self.verschleierungsVerdacht = false
        self.modus = .heuristisch
        self.hinweise = []
        self.analyseDatum = Date()
        self.frequenzBandAnteile = []
        self.frequenzBandGrenzen = []
        self.segmentZeiten = []
        self.segmentStaerken = []
        self.spektrumFrequenzen = []
        self.spektrumWerte = []
        self.fingerprintWerte = []
        self.grundlinieWerte = []
        self.ssmMatrix = []
    }
}

// MARK: - Trainings-Probe
struct TrainingsProbe: Identifiable, Codable {
    let id: UUID
    var dateiPfad: String
    var dateiName: String
    var label: TrainingsLabel
    var merkmalsvektor: [Double]   // 128-dimensional
    var dateiSignatur: String      // MD5-Hash für Deduplizierung
    var hinzugefuegtAm: Date
    
    enum TrainingsLabel: Int, Codable {
        case echt = 0
        case ki = 1
        
        var bezeichnung: String {
            switch self { case .echt: return "Echte Musik"; case .ki: return "KI-Musik" }
        }
    }
}

// MARK: - ML Modell
struct LogistischesModell: Codable {
    var gewichte: [Double]   // 128 Gewichte + 1 Bias
    var bias: Double
    var trainingsZeit: Date
    var anzahlBeispiele: Int
    var kreuzvalidierungsGenauigkeit: Double
    
    func vorhersage(merkmale: [Double]) -> Double {
        guard merkmale.count == gewichte.count else { return 0.5 }
        var summe = bias
        for i in 0..<merkmale.count {
            summe += gewichte[i] * merkmale[i]
        }
        return sigmoid(summe)
    }
    
    private func sigmoid(_ x: Double) -> Double {
        if x >= 0 { return 1.0 / (1.0 + exp(-x)) }
        let z = exp(x); return z / (1.0 + z)
    }
}
